### DISCUSSION TITLE
SourceMap header or sourceMappingUrl comment for static sites?

### DISCUSSION THREAD

**Discussion by bmulholland** [nuxt-community/sentry-module#369] [Q&A] [answered]
I have configured publishRelease, and that is working: the release is pushed to Sentry, and source code (and maps) are uploaded to the project. Nevertheless, the mapping isn't working: Sentry tells me "Source code was not found"

Now, looking at [their troubleshooting tips](https://docs.sentry.io/platforms/javascript/sourcemaps/troubleshooting_js/), they suggest checking for `sourceMappingUrl` - and that isn't set. Looking at [the sentry-webpack-plugin docs](https://github.com/getsentry/sentry-webpack-plugin#options), the default is to use a SourceMap header - and that header is not set on my production server. Likewise, sourceMappingUrl comments are not added. `sentry-webpack-plugin` includes notes on how to disable the SourceMap header -- assuming `sourceMappingUrl` is set in all the files -- but there's no mention of how to do that anywhere in this plugin, in sentry-webpack-plugin, _or_ in sentry-cli.

Questions:
* I feel like I must be missing something? How am I the first person to run into this?
* Is the expectation of this plugin that nuxt is serving the files, and thus setting the SourceMap header?
* If I'm hosting my site statically, do I need to set the sourceMappingUrl somehow? Or is there some clever way to server the `SourceMap` header?
* How do I get the sourceMappingUrl added to the generated assets?

Configuration (yes, all env vars are properly set, as confirmed by the source files making it to Sentry fine):
```
    sentry: {
      dsn: process.env.SENTRY_DSN,
      // Enable performance tracing
      tracing: {
        // 1.0 = 100% of requests will be sampled, so 0.2 = 20%
        tracesSampleRate: 0.2,
      },
      // Send source maps to Sentry for better error messages
      publishRelease: {
        authToken: process.env.SENTRY_AUTH_TOKEN,
        org: process.env.SENTRY_ORG,
        project: process.env.SENTRY_PROJECT,
      },
      // https://sentry.nuxtjs.org/sentry/options/#sourcemapstyle
      // most people, that should be a better option but due to it being a
      // breaking change, it won't be set as the default until next major
      // release
      sourceMapStyle: "hidden-source-map",
    },
```

**Comment by rchl** [comment, marked as answer]
Documentation suggests setting `sourceMapStyle` to `hidden-source-map` but I'm not actually sure if that works correctly with sentry.

Can you try removing the `sourceMapStyle` option so that it uses `source-map` instead. Then the source files should include `//# sourceMappingURL` comment at the end.

**Reply by bmulholland** [reply]
In the kindest tone possible: why do the docs recommend something that isn't confirmed working for the happy-path?

In any case, Sentry support let me know that the problem appears to be that the files are uploaded to Sentry with a relative path "~/_nuxt/2969e7f.js", while issues are reported to Sentry using an absolute path: "File@https://example.com/_nuxt/2969e7f.js"

Looking through the source of this project, [I noticed it constructs paths](https://github.com/nuxt-community/sentry-module/blob/607511a4e4c6000fbea78e811249839d4a1f7272/lib/core/hooks.js#L193) using [nuxt's build publicPath](https://nuxtjs.org/docs/configuration-glossary/configuration-build#publicpath), so I tried setting that. It was closer, now "~/https:/example.com/2969e7f.js" - but still not right:
* starts with relative "~/" even though it's an absolute path now
* missing "File@" prefix
* should be two slashes after "https:"

At this point, I note the code appears to be wrong. Specifically these lines:
https://github.com/nuxt-community/sentry-module/blob/607511a4e4c6000fbea78e811249839d4a1f7272/lib/core/hooks.js#L193:L194

1. As documented, [router.base is a relative path](https://nuxtjs.org/docs/configuration-glossary/configuration-router/#base) ([confirmation from vue-router](https://github.com/vuejs/vue-router/issues/3372#issuecomment-727024198)), while [publicPath is an absolute path](https://nuxtjs.org/docs/configuration-glossary/configuration-build#publicpath). So why is the relative path first in the join? That's backwards.
2. `posix` is for posix paths, but we're constructing a URL. That's probably why the "//" is being collapsed into a single slash
2. Since the relative path is first, and is blank for me, the path from a path join will always start with a "/". That's where the "~" is coming from. (And actually, unless misconfigured, the absolute-path case is currently inaccessible.)
3. The last thing is the "File@" prefix - I'm not sure whether the [RewriteFrames option](https://docs.sentry.io/platforms/javascript/configuration/integrations/plugin/#rewriteframes) should be used to remove it, or the prefix should be added as part of the URL construction above. Since both parts are within control of this plugin, though, I'd expect it to be taken care of.

At this point, the fix seems to be:
1. Clarify the documentation to say that `build.publicPath` is required to be set for source maps to work when hosted from a CDN.
2. Update the docs to mention urlPrefix, which is currently completely undocumented.
3. Update the urlPrefix to correctly construct a client-facing path: if public path is set, use `new URL` instead of `posix`, otherwise use only `router.base`.

**Reply by bmulholland** [reply]
I tried checking the code out to make a PR but I can't even run a `yarn install` (see below) - so I'll just set the urlPrefix, which will work for me, anyway.

```
error /Users/bmulholland/dev/oss/sentry-module/node_modules/playwright-chromium: Command failed.
Exit code: 1
Command: node install.js
Arguments:
Directory: /Users/bmulholland/dev/oss/sentry-module/node_modules/playwright-chromium
Output:
/Users/bmulholland/dev/oss/sentry-module/node_modules/proper-lockfile/lib/lockfile.js:68
                return callback(Object.assign(new Error('Lock file is already being held'), { code: 'ELOCKED', file }));
                                              ^

Error: Lock file is already being held
    at /Users/bmulholland/dev/oss/sentry-module/node_modules/proper-lockfile/lib/lockfile.js:68:47
    at callback (/Users/bmulholland/dev/oss/sentry-module/node_modules/graceful-fs/polyfills.js:299:20)
    at FSReqCallback.oncomplete (node:fs:199:5) {
  code: 'ELOCKED',
  file: '/Users/bmulholland/Library/Caches/ms-playwright'
```

**Reply by rchl** [reply]
> In the kindest tone possible: why do the docs recommend something that isn't confirmed working for the happy-path?

Honest mistake. I think I've set it based on recommendations but forgot to change after realizing that it had some issues.

> In any case, Sentry support let me know that the problem appears to be that the files are uploaded to Sentry with a relative path "~/_nuxt/2969e7f.js", while issues are reported to Sentry using an absolute path: "File@example.com/_nuxt/2969e7f.js"

Can you post some example, raw, minified stack trace?

Here I get stack traces like:

```
TypeError: Cannot read properties of undefined (reading 'push')
  at ? (/_nuxt/ddb2d19.js:1:147685)
  at v(/_nuxt/939375f.js:2:10776)
  at Generator._invoke(/_nuxt/939375f.js:2:10564)
  at Generator.throw(/_nuxt/939375f.js:2:11205)
  at r(/_nuxt/939375f.js:2:161)
  at h(/_nuxt/939375f.js:2:400)
```

And source maps are uploaded with paths like `~/_nuxt/0004c08.js` and that works.

Are you saying that your errors have full paths in those original stack traces? If that's the case then there is something weird about your project that is not accounted for. I'd like to know more details to be able to understand that.

The logic for determining the prefix maybe doesn't account for that case correctly but it should work with basic setup.

> As documented, router.base is a relative path (confirmation from vue-router), while publicPath is an absolute path. So why is the relative path first in the join? That's backwards.

`publicPath` is by default `/_nuxt/` so it can be either relative or absolute.

> I tried checking the code out to make a PR but I can't even run a yarn install (see below) - so I'll just set the urlPrefix, which will work for me, anyway.

Works here. Just tried after removing `node_modules`.

**Reply by rchl** [reply]
> Update the urlPrefix to correctly construct a client-facing path: if public path is set, use new URL instead of posix, otherwise use only router.base.

Rather, use `publicPath` if absolute, otherwise construct as it is done currently (base path first then public path).

**Reply by bmulholland** [reply]
> Honest mistake.

These things happen :)

Yes, the errors have full paths, and this happened _before_ I made nuxt aware of the URL via configuration. It's interesting that the base case should be passing through relative paths but was coming through as absolute for me - I don't _think_ the configuration is in any way unusual in this respect. Where in the code are these paths constructed?

Here's an example raw stack trace with the full (anonymized) URL included:

```
TypeError: this.comparison is null
  at revised(**File@https://example.com/_nuxt/2969e7f.js**:1:9933)
  at get(./node_modules/vue/dist/vue.runtime.esm.js:4495:25)
  at evaluate(./node_modules/vue/dist/vue.runtime.esm.js:4597:21)
  at Pn/<(./node_modules/vue/dist/vue.runtime.esm.js:4851:17)
  at apply(./components/dumb/FileComparisonCard.vue?8663:3:615)
  at scopedSlotFn(./node_modules/vue/dist/vue.runtime.esm.js:2605:37)
  at renderSlot(./node_modules/vue/dist/vue.runtime.esm.js:2701:7)
  at call(./components/dumb/html/Accordion.vue?9e35:1:396)
  at _render(./node_modules/vue/dist/vue.runtime.esm.js:3569:22)
  at call(./node_modules/vue/dist/vue.runtime.esm.js:4081:21)
  at get(./node_modules/vue/dist/vue.runtime.esm.js:4495:25)
  at run(./node_modules/vue/dist/vue.runtime.esm.js:4570:22)
  at flushSchedulerQueue(./node_modules/vue/dist/vue.runtime.esm.js:4326:13)
  at i(./node_modules/vue/dist/vue.runtime.esm.js:1989:12)
  at flushCallbacks(./node_modules/vue/dist/vue.runtime.esm.js:1915:12)
```

On yarn install playwright - I'm on an M1 MBP, which is probably the issue. Unfortunately I don't have any clever ways around the problem, or even a way to diagnose.

**Reply by rchl** [reply]
This is an error from Chrome? I don't get how it would be so different from what I'm getting...

This needs to be figured out before we do any changes. I can imagine that it would be like that when using absolute `publicPath` but not otherwise. Perhaps you have some module that does change some Nuxt Config values for you?

> On yarn install playwright - I'm on an M1 MBP, which is probably the issue.

I'm also on M1 MBP on Node 14.

**Reply by bmulholland** [reply]
Firefox and Safari both reported like that, possibly Chrome but I don't have a trace handy.

I've now set the urlPrefix and source maps are uploaded to Sentry with a URL path that matches the raw trace above. But now the traces have switched to relative! So my workaround doesn't even work.

Example:

```
TypeError: t.data is undefined
  at result(/df19766.js:1:46725)
  at value(/b3b961a.js:2:601177)
  at value(/b3b961a.js:2:601461)
  at k(/b3b961a.js:2:820582)
  at S(/b3b961a.js:2:821039)
  at value(/b3b961a.js:2:821657)
  at za/<(/b3b961a.js:2:394870)
  at za(/b3b961a.js:2:394838)
  at r(/b3b961a.js:2:393910)
  at o(/b3b961a.js:2:408119)
```

**Reply by bmulholland** [reply]
I can do some debugging but need to know where to look. Can you point me to where these stacktrace paths are generated?

**Reply by rchl** [reply]
The browser generates the stack trace initially and then it's captured by Sentry.

Nuxt doesn't do any post-processing for client-side error I believe.

Sentry might be doing post-processing but I'm not sure if on the client, before sending or after the crash is sent to Sentry.

Though it appears to me now that the stack you've posted in https://github.com/nuxt-community/sentry-module/discussions/369#discussioncomment-1686529 is already processed since it contains `node_modules` paths rather just the paths to the bundle.

Can you post the stack trace from the issue with those buttons selected?

![Screenshot 2021-11-23 at 12 20 39](https://user-images.githubusercontent.com/153197/143015613-d3ba16b3-5def-41f9-a7a3-85fc79f138d8.png)

The `minified` view is the original stack trace as reported by the browser, I believe. The `original` is already processed by Sentry (or source mapped).

**Reply by bmulholland** [reply]
Minified looks similar:

```
TypeError: this.comparison is null
  at revised(File@https://example.com/_nuxt/2969e7f.js:1:9933)
  at kn.prototype.get(/_nuxt/e252db4.js:2:27382)
  at kn.prototype.evaluate(/_nuxt/e252db4.js:2:28522)
  at Pn/<(/_nuxt/e252db4.js:2:30411)
  at fn(/_nuxt/2969e7f.js:1:11146)
  at r(/_nuxt/e252db4.js:2:15927)
  at Ie(/_nuxt/e252db4.js:2:16765)
```

I'm going to try again without the URL prefix set and see what values I get on both sides.

**Reply by bmulholland** [reply]
After a lot of trial and error, all appears working -- with the default config, no less -- now.

I'm pretty sure the only real diff since I started working on it is removing [nuxt/emotion](https://github.com/nuxt-community/emotion-module) - which we weren't even using. I removed it as suspect because it claims source maps as one of its features, and that seems to have been the case...

**Reply by bmulholland** [reply]
Okay, I tried re-adding emotion and still everything is working. So at the end, I've changed literally nothing and it's now working where it wasn't before?

I'm out of ideas for what the issue could have been, or where those `File@`'s were coming from, so I guess we'll never know...

Thanks again for all your help, that was a weird one and I couldn't have done it without you.

**Reply by bmulholland** [reply]
Oh, to try to give some value back: I am using hidden-source-maps and that seems to be working well.
